### PR TITLE
Feat/add syslog capabilities 4 opentelemetry collector

### PR DESCRIPTION
**Pull request by phozzy**
## Description of changes

Solves #326067

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by arianvp** [comment]
I think adding `CAP_SYSLOG` is not the right thing. that's for logging *to* syslog. Not reading from the journal.   You need to add the DynamicUser to the `systemd-journald` and/or `adm` groups. We had the same issue with Grafana Alloy  (a fork/ competitor of otel-collector): https://github.com/NixOS/nixpkgs/pull/318306

**Comment by arianvp** [comment]
Tagging flokli as he solved the exact same issue for grafana-alloy

**Comment by phozzy** [comment]
> I think adding `CAP_SYSLOG` is not the right thing. that's for logging _to_ syslog. Not reading from the journal. You need to add the DynamicUser to the `systemd-journald` and/or `adm` groups. We had the same issue with Grafana Alloy (a fork/ competitor of otel-collector): #318306

@arianvp Oh... That's a good suggestion. Was not certain that `SupplementaryGroups` can do the trick. Will change my PR then.

**Comment by arianvp** [comment]
Congrats on your first PR by the way. If you could be so kind could you please format your commit message according to our contributing guide?   After that I can give this a shot and test it out and then we can merge it. https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md

**Comment by phozzy** [comment]
@arianvp is it good enough now?

File: nixos/modules/services/monitoring/opentelemetry-collector.nix [hunk 1]
```diff
@@ -48,21 +49,27 @@ in {
...
+            else cfg.configFile;
+        in
+        {
+          ExecStart = "${getExe opentelemetry-collector} --config=file:${conf}";
+          DynamicUser = true;
+          Restart = "always";
+          ProtectSystem = "full";
+          DevicePolicy = "closed";
+          NoNewPrivileges = true;
+          WorkingDirectory = "/var/lib/opentelemetry-collector";
```
Comment (aanderse) [inline nixos/modules/services/monitoring/opentelemetry-collector.nix:66]: not related to this PR at all but shouldn't that be

```suggestion
          WorkingDirectory = "%S/opentelemetry-collector";
```

instead?

**Review by aanderse — COMMENTED** [review]
any chance we can either drop the formatting changes or put them into a separate commit? makes this change really hard to track... 😅 

but yeah this PR looks right to me 👍

File: nixos/modules/services/monitoring/opentelemetry-collector.nix (hunk 1 above)
Comment (phozzy) [reply nixos/modules/services/monitoring/opentelemetry-collector.nix:66]: It used to be on a separate commit, but I was not sure how to keep it according to the contribution guide. So, I squashed it to a single commit.

Comment (aanderse) [reply nixos/modules/services/monitoring/opentelemetry-collector.nix:66]: i'm not sure i follow... it is difficult to spot your change now

i think you should either drop the formatting or put in separate commit - what part of contributing guidelines are you referencing?

Comment (phozzy) [reply nixos/modules/services/monitoring/opentelemetry-collector.nix]: Now there are 3 commits:
1. Fix formatting
2. Add a supplementary group
3. Avoid fixed path

Comment (aanderse) [reply nixos/modules/services/monitoring/opentelemetry-collector.nix]: wonderful!

**Pull request merged**